cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfRobotComponent.cpp

```diff
@@ -0,0 +1,315 @@
+        // a joint once the parent link and child link have already been created. But the UUrdfLinkComponent::initialize(...)
+        // method is designed to be called as soon as a parent link has been created, and before a child link has been created.
+        // There are several alternative approaches that would also work here, but we think this approach is the cleanest as
+        // measured by the simplicity of our URDF parsing code, the simplicity of our UrdfRobotDesc data structure, and the
+        // simplicity of our recursive code for creating the Unreal component hierarchy.
+
+        SP_ASSERT(!Std::containsSubstring(child_joint_desc->name_, "."));
+        auto child_joint_component = NewObject<UUrdfJointComponent>(this, Unreal::toFName(child_joint_desc->name_));
+        SP_ASSERT(child_joint_component);
+        child_joint_component->SetupAttachment(parent_link_component);
+        child_joint_component->RegisterComponent();
+        // It is very important to setup attachments, and register components before we create more child components.
+        // If this order is reversed and we end up creating more child components before registering the parent component,
+        // the physics simulation will be unstable.
+        child_joint_component->initialize(child_joint_desc, parent_link_component, child_link_component);
+        JointComponents.Add(child_joint_component);
+
+        initialize(child_link_desc, child_link_component);
+    }
+}
+
+void UUrdfRobotComponent::initializeDeferred()
+{
+    // Cache components in maps so we can refer to them by name. We need to do this in BeginPlay because after pressing play
+    // in the editor, the Unreal Engine creates a new replica object for each object in the World Outliner. For any object
+    // that we spawned in the editor, we need to re-compute any local state that isn't visible to the Unreal reflection system.
+    // We can't re-compute this local state in the constructor because LinkComponents and JointComponents have not been updated
+    // by the the Unreal reflection system yet. So we need to do it in BeginPlay.
+
+    for (auto link_component : LinkComponents) {
+        link_components_[Unreal::toStdString(link_component->GetName())] = link_component;
+    }
+
+    for (auto joint_component : JointComponents) {
+        joint_components_[Unreal::toStdString(joint_component->GetName())] = joint_component;
+    }
+
+    // We cache root link component's reference to a local variable root_link_component_ in initalize().
+    // In Editor mode, this root_link_component_ gets broken for the same reason mentioned above, hence
+    // we need to cache it again here.
+    if (!root_link_component_ && LinkComponents.Num()) {
+        root_link_component_ = LinkComponents[0];
+    }
+    SP_ASSERT(root_link_component_);
+
+    for (auto joint_component : JointComponents) {
```

Comment:
Don't call `joint_component->initializeDeferred()` here because it should be the joint's responsibility to call its own `initializeDeferred()` method.

Reply:
Great!